crates/mcp-server/src/repository_inference.rs

```
@@ -0,0 +1,183 @@
+//! Repository inference from current working directory
+//!
+//! Matches CWD against indexed repositories to automatically select
+//! the appropriate repository for search queries.
+
+use crate::error::{McpError, Result};
+use std::path::Path;
+use uuid::Uuid;
+
+/// Repository info from database
+#[derive(Debug, Clone)]
+pub struct IndexedRepository {
+    pub id: Uuid,
+    pub name: String,
+    pub path: String,
+}
+
+/// Infer repository from current working directory
+///
+/// Walks up the directory tree from CWD looking for a match
+/// against indexed repository paths.
+pub fn infer_repository_from_cwd(
+    cwd: &Path,
+    indexed_repos: &[IndexedRepository],
+) -> Result<Vec<String>> {
+    if indexed_repos.is_empty() {
+        return Err(McpError::RepositoryInference(
+            "No indexed repositories found. Run 'codesearch index' first.".to_string(),
+        ));
+    }
+
+    // Try to find a repository that contains the CWD
+    let cwd_canonical = cwd.canonicalize().unwrap_or_else(|_| cwd.to_path_buf());
```

If `canonicalize()` fails, the fallback silently uses the uncanonicalized path. This could cause false negatives in repository matching. Consider logging a warning when canonicalization fails.